daemon/daemon.go

```diff
@@ -511,6 +530,32 @@ var (
 	rebootMaxTentatives    = 3
 )
 
+func (d *Daemon) updateMaintenanceFile(rst state.RestartType) error {
+	// for unset restart, just remove the maintenance.json file
+	if rst == state.RestartUnset {
+		err := os.Remove(dirs.SnapdMaintenanceFile)
+		// only return err if the error was something other than the file not
+		// existing
+		if err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		return nil
+	}
+
+	// otherwise marshal and write it out appropriately
+	b, err := json.Marshal(maintenanceForRestartType(rst))
+	if err != nil {
+		return err
+	}
+
+	err = osutil.AtomicWrite(dirs.SnapdMaintenanceFile, bytes.NewBuffer(b), 0644, 0)
```

Comment:
(nitpick^4) could be a single line: `if err := ...` or even just a `return osutil.AtomicWrite(...`